Add various Vertica connection parameters

darkag opened the pull request:
Give the possibility to pass vertica connection parameters using extra. Especially it gives the possibility to enable load balancing (Vertica requires that load balancing is enabled client and server side)

darkag committed:
first try parameters

darkag committed:
Add binding to all options

Add binding from extra to vertica connection parameters

Adaverse commented on airflow/providers/vertica/hooks/vertica.py:
```diff
@@ -45,6 +45,51 @@ def get_conn(self) -> connect:
         else:
             conn_config["port"] = int(conn.port)
+            
+        if conn.extra_dejson.get("connection_load_balance", False):
```
Can we fetch `extra` dict by calling `conn.extra_dejson` once rather than calling it repeatedly? It makes it look very wordy

darkag replied:
Yes, I based my code on what it was done for the mysql hook, I didn't realize that extra_dejson was property that load json every time you call it.

Adaverse replied:
It's not that big of a performance concern since it would happen only while adding the connection (which is one time process). But the code looks very wordy.

Adaverse commented on airflow/providers/vertica/hooks/vertica.py:
```diff
@@ -45,6 +45,51 @@ def get_conn(self) -> connect:
+        if conn.extra_dejson.get("kerberos_service_name", False):
+            conn_config["kerberos_service_name"] = conn.extra_dejson["kerberos_service_name"]
+
+        if conn.extra_dejson.get("log_level", False):
```
is `False` the right default value for `log_level`. From what I can see from  [here](https://github.com/vertica/vertica-python#logging), it seems to be _logging.WARNING_.

The same question for `log_path` and many others...

darkag replied:
the .get(<option_name>, False) is not used for declare a default value but for test the presence of the option key in extra dictionnary.

Adaverse replied:
mb seems like i presumed something i shouldn't. Looks good here!

darkag committed:
Update vertica.py

Simplification of code, better handling of log level parameter, correction for "request_complex_types" which is true by default (it would have been impossible to pass false with the previous version)

darkag committed:
Fix to pass test

Fix useless spaces, pass options only if present in extra, loop over options to reduce code size

Adaverse commented on airflow/providers/vertica/hooks/vertica.py:
```diff
@@ -46,5 +46,41 @@ def get_conn(self) -> connect:
 
+        boolOptions = ["connection_load_balance", "binary_transfer", "disable_copy_local", "request_complex_types", "use_prepared_statements"]
+        stdOptions = ["session_label", "backup_server_node", "kerberos_host_name", "kerberos_service_name", "log_path", "unicode_error", "workload", "ssl"]
+        conn_extra = conn.extra_dejson
```
Any reason for going for a camel case instead of snake? It should be latter.

darkag replied:
my bad, the habit of coding in javascript/c#. I've fixed it

darkag committed:
fix naming convention

Commit:
add tests and doc

potiuk commented on airflow/providers/vertica/hooks/vertica.py:
```diff
@@ -46,5 +46,41 @@ def get_conn(self) -> connect:
             conn_config["port"] = int(conn.port)
 
+        bool_options = ["connection_load_balance", "binary_transfer", "disable_copy_local", "request_complex_types", "use_prepared_statements"]
+        std_options = ["session_label", "backup_server_node", "kerberos_host_name", "kerberos_service_name", "log_path", "unicode_error", "workload", "ssl"]
```
I carefeully looked at the parameters passed, and I think being able to set  log_path from extra is potentially dangerous and we've learned that we should carefully look at the parameters that you pass from connection definition. Users who have the rights to edit the connection should not have some additional priviledges, and there is a subtle exception here in "log_path". 

Enabling log_path to be set "freely" opens the doors to a subtle security issue, where you could override some files by setting the log path when you configure the connecion to point to a file that you want to override. I personally think that we should not allow log_path setting at all and leave it with the default "vertica_python.log" in default directory. 

But maybe there are other proposals?

darkag replied:
I've made a try setting log_level without setting log_path, default directory turns to be / when runned by scheduler which make exexution fail since it doesn't have root rights. But, since vertica logging is visible in dag run logs, the log file may not be really useful, so I propose to remove the log_path parameter (in order to remove potential security issue) and pass /dev/null as log path when log_level is defined. Doing this we keep the possibility to have traces from vertica client but without the inconvient to have a log file who will contains the same informations than the dag run logs.

If it's ok for you I'll make the change in code, if not I prefer to remove the logging parameters

darkag replied:
Finnally I set log_path to None when log_level is set, it's prevent generation of log file and is cleaner than /dev/null

potiuk replied:
Looks good. Just remove it also from examples :).

darkag replied:
done

Commit:
Remove log_path from Extra

Commit:
remove log_path from example

phanikumv commented on tests/providers/vertica/hooks/test_vertica.py:
```diff
@@ -47,6 +47,57 @@ def test_get_conn(self, mock_connect):
 
+    @patch("airflow.providers.vertica.hooks.vertica.connect")
+    def test_get_conn_extra_parameters_no_cast(self, mock_connect):
+        extra_dict = self.connection.extra_dejson
```
what is the purpose of this test? Can you please add docstring here. It will be helpful for everyone

phanikumv commented on tests/providers/vertica/hooks/test_vertica.py:
```diff
@@ -47,6 +47,57 @@ def test_get_conn(self, mock_connect):
+
+    @patch("airflow.providers.vertica.hooks.vertica.connect")
+    def test_get_conn_extra_parameters_cast(self, mock_connect):
+        import logging
```
what is the purpose of this test? Can you please add docstring here. It will be helpful for everyone

darkag replied:
I've added a docstring, this test is for parameters that can be passed as string or as int/bool, the first test test them as int/bool. The second as string. Since I don't really know how the mocking works I was not sure that I can make both test in one function and maybe it's possible to chain both case in on test like this :

       std_options = ["session_label", "kerberos_host_name", "kerberos_service_name", "unicode_error", "workload", "ssl"]
        for so in std_options:
            extra_dict.update({ so: so})
        bck_server_node = ["1.2.3.4", "4.3.2.1"]
        conn_timeout = 30
        log_lvl = 40
        extra_dict.update({"backup_server_node":bck_server_node})
        extra_dict.update({"connection_timeout":conn_timeout})
        extra_dict.update({"log_level":log_lvl})

        self.db_hook.get_conn()
        assert mock_connect.call_count == 1
        args, kwargs = mock_connect.call_args
        assert args == ()
        for bo in bool_options:
            assert kwargs[bo] == True
        assert kwargs["request_complex_types"] == False
        for so in std_options:
            assert kwargs[so] == so
        assert bck_server_node[0] in kwargs["backup_server_node"]
        assert bck_server_node[1] in kwargs["backup_server_node"]
        assert kwargs["connection_timeout"] == conn_timeout
        assert kwargs["log_level"] == log_lvl
        assert kwargs["log_path"] is None

        for bo in bool_options:
            extra_dict.update({ bo: "True"})
        extra_dict.update({ "request_complex_types": "False"})
        extra_dict.update({"log_level":"Error"})
        self.db_hook.get_conn()
        assert mock_connect.call_count == 1 # 2 ??
        args, kwargs = mock_connect.call_args
        assert args == ()
        for bo in bool_options:
            assert kwargs[bo] == True
        assert kwargs["request_complex_types"] == False
        assert kwargs["log_level"] == logging.ERROR
        assert kwargs["log_path"] is None

Commit:
add docstring for test_get_conn_extra_parameters_cast

Commit:
docstring

Commit:
Fix error in tests and typo for static checks

darkag committed:
multiples fixes (doc, test, parsing)
